File: coral-go/internal/server/routes/webhooks.go
package routes

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/cdknorow/coral/internal/config"
	"github.com/cdknorow/coral/internal/httputil"
	"github.com/cdknorow/coral/internal/store"
)

// WebhooksHandler handles webhook configuration and delivery endpoints.
type WebhooksHandler struct {
	ws  *store.WebhookStore
	cfg *config.Config
}

func NewWebhooksHandler(db *store.DB, cfg *config.Config) *WebhooksHandler {
	return &WebhooksHandler{
		ws:  store.NewWebhookStore(db),
		cfg: cfg,
	}
}

// ListWebhooks returns all webhook configurations.
// GET /api/webhooks
func (h *WebhooksHandler) ListWebhooks(w http.ResponseWriter, r *http.Request) {
	configs, err := h.ws.ListWebhookConfigs(r.Context(), false)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, emptyIfNil(configs))
}

// CreateWebhook creates a new webhook.
// POST /api/webhooks
func (h *WebhooksHandler) CreateWebhook(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name        string  `json:"name"`
		Platform    string  `json:"platform"`
		URL         string  `json:"url"`
		AgentFilter *string `json:"agent_filter"`
	}
	if err := decodeJSON(r, &body); err != nil {
		errBadRequest(w, "invalid JSON")
		return
	}
	if body.Name == "" || body.URL == "" {
		errBadRequest(w, "name and url are required")
		return
	}
	if body.Platform == "" {
		body.Platform = "generic"
	}
	wh, err := h.ws.CreateWebhookConfig(r.Context(), body.Name, body.Platform, body.URL, body.AgentFilter)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, wh)
}

// UpdateWebhook updates a webhook configuration.
// PATCH /api/webhooks/{webhookID}
func (h *WebhooksHandler) UpdateWebhook(w http.ResponseWriter, r *http.Request) {
	whID, _ := strconv.ParseInt(chi.URLParam(r, "webhookID"), 10, 64)
	var fields map[string]interface{}
	if err := decodeJSON(r, &fields); err != nil {
		errBadRequest(w, "invalid JSON")
		return
	}
	if err := h.ws.UpdateWebhookConfig(r.Context(), whID, fields); err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"ok": true})
}

// DeleteWebhook deletes a webhook and its delivery history.
// DELETE /api/webhooks/{webhookID}
func (h *WebhooksHandler) DeleteWebhook(w http.ResponseWriter, r *http.Request) {
	whID, _ := strconv.ParseInt(chi.URLParam(r, "webhookID"), 10, 64)
	if err := h.ws.DeleteWebhookConfig(r.Context(), whID); err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"ok": true})
}

// TestWebhook sends a test notification immediately via direct delivery.
// POST /api/webhooks/{webhookID}/test
func (h *WebhooksHandler) TestWebhook(w http.ResponseWriter, r *http.Request) {
	whID, _ := strconv.ParseInt(chi.URLParam(r, "webhookID"), 10, 64)
	cfg, err := h.ws.GetWebhookConfig(r.Context(), whID)
	if err != nil || cfg == nil {
		errNotFound(w, "Webhook not found")
		return
	}
	// SSRF protection: validate webhook URL doesn't target internal networks
	if _, err := httputil.ResolveAndValidateURL(cfg.URL); err != nil {
		errBadRequest(w, fmt.Sprintf("webhook URL blocked: %v", err))
		return
	}
	delivery, err := h.ws.CreateWebhookDelivery(
		r.Context(), whID, "coral-test", "needs_input",
		"Test notification from Coral dashboard", nil,
	)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	// Deliver immediately inline (same as Python's deliver_now)
	go h.deliverTestWebhook(cfg, delivery)
	// Return the delivery record
	writeJSON(w, http.StatusOK, delivery)
}

func (h *WebhooksHandler) deliverTestWebhook(cfg *store.WebhookConfig, delivery *store.WebhookDelivery) {
	ctx := context.Background()
	payload := map[string]interface{}{
		"agent_name": delivery.AgentName,
		"session_id": delivery.SessionID,
		"event_type": delivery.EventType,
		"summary":    delivery.EventSummary,
		"timestamp":  delivery.CreatedAt,
		"source":     "coral",
	}
	body, _ := json.Marshal(payload)

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequestWithContext(ctx, "POST", cfg.URL, bytes.NewReader(body))
	if err != nil {
		errMsg := err.Error()
		h.ws.MarkWebhookDelivery(ctx, delivery.ID, "failed", nil, &errMsg, nil, nil)
		return
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		errMsg := err.Error()
		if len(errMsg) > 200 {
			errMsg = errMsg[:200]
		}
		h.ws.MarkWebhookDelivery(ctx, delivery.ID, "failed", nil, &errMsg, nil, nil)
		return
	}
	defer resp.Body.Close()

	status := resp.StatusCode
	if status >= 200 && status < 300 {
		h.ws.MarkWebhookDelivery(ctx, delivery.ID, "delivered", &status, nil, nil, nil)
	} else {
		errMsg := fmt.Sprintf("HTTP %d", status)
		h.ws.MarkWebhookDelivery(ctx, delivery.ID, "failed", &status, &errMsg, nil, nil)
	}
}

// ListDeliveries returns delivery history for a webhook.
// GET /api/webhooks/{webhookID}/deliveries
func (h *WebhooksHandler) ListDeliveries(w http.ResponseWriter, r *http.Request) {
	whID, _ := strconv.ParseInt(chi.URLParam(r, "webhookID"), 10, 64)
	limit := queryInt(r, "limit", 50)
	if limit > 200 {
		limit = 200
	}
	deliveries, err := h.ws.ListWebhookDeliveries(r.Context(), whID, limit)
	if err != nil {
		errInternalServer(w, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, emptyIfNil(deliveries))
}

File: coral-python/coral/api/webhooks.py
"""API routes for webhook configuration and delivery history."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING
from urllib.parse import urlparse

from fastapi import APIRouter, Query

if TYPE_CHECKING:
    from coral.store import CoralStore

log = logging.getLogger(__name__)

router = APIRouter()

# Module-level dependencies, set by web_server.py during app setup
store: CoralStore = None  # type: ignore[assignment]
_app = None  # Set by web_server for accessing app.state


VALID_PLATFORMS = {"slack", "discord", "generic"}


def _validate_url(url: str, platform: str) -> str | None:
    """Return error message if URL is invalid, else None."""
    try:
        parsed = urlparse(url)
    except Exception:
        return "Invalid URL"
    if parsed.scheme not in ("http", "https"):
        return "URL must use http or https"
    if parsed.scheme == "http" and parsed.hostname not in ("localhost", "127.0.0.1"):
        return "HTTP (non-HTTPS) is only allowed for localhost"
    if not parsed.netloc:
        return "URL must have a hostname"
    return None


@router.get("/api/webhooks")
async def list_webhooks():
    """List all webhook configurations."""
    return await store.list_webhook_configs()


@router.post("/api/webhooks")
async def create_webhook(body: dict):
    """Create a new webhook configuration."""
    name = body.get("name", "").strip()
    platform = body.get("platform", "generic").strip()
    url = body.get("url", "").strip()
    if not name or not url:
        return {"error": "name and url are required"}
    if platform not in VALID_PLATFORMS:
        return {"error": f"platform must be one of: {', '.join(sorted(VALID_PLATFORMS))}"}
    url_error = _validate_url(url, platform)
    if url_error:
        return {"error": url_error}
    return await store.create_webhook_config(
        name=name,
        platform=platform,
        url=url,
        agent_filter=body.get("agent_filter") or None,
    )


@router.patch("/api/webhooks/{webhook_id}")
async def update_webhook(webhook_id: int, body: dict):
    """Update fields on a webhook configuration."""
    if "url" in body:
        url_error = _validate_url(body["url"], body.get("platform", "generic"))
        if url_error:
            return {"error": url_error}
    await store.update_webhook_config(webhook_id, **body)
    return {"ok": True}


@router.delete("/api/webhooks/{webhook_id}")
async def delete_webhook(webhook_id: int):
    """Delete a webhook configuration and all its delivery history."""
    await store.delete_webhook_config(webhook_id)
    return {"ok": True}


@router.post("/api/webhooks/{webhook_id}/test")
async def test_webhook(webhook_id: int):
    """Send a test notification immediately via direct delivery."""
    cfg = await store.get_webhook_config(webhook_id)
    if not cfg:
        return {"error": "Webhook not found"}
    delivery = await store.create_webhook_delivery(
        webhook_id=webhook_id,
        agent_name="coral-test",
        session_id=None,
        event_type="needs_input",
        event_summary="Test notification from Coral dashboard",
    )
    dispatcher = getattr(_app.state, "webhook_dispatcher", None) if _app else None
    if dispatcher:
        await dispatcher.deliver_now(delivery)
    # Re-fetch to get updated status after delivery attempt
    deliveries = await store.list_webhook_deliveries(webhook_id, limit=1)
    return deliveries[0] if deliveries else {"ok": True}


@router.get("/api/webhooks/{webhook_id}/deliveries")
async def list_deliveries(
    webhook_id: int, limit: int = Query(50, ge=1, le=200)
):
    """Get recent delivery history for a webhook."""
    return await store.list_webhook_deliveries(webhook_id, limit=limit)
